### PR TITLE
[Relay][SimplifyInference] Express Softmax as sequence of Relay ops

### PR DESCRIPTION
**Pull request by anijain2305**
Discuss - https://discuss.tvm.ai/t/softmax-sequence-of-relay-ops/5686

@soiferj @yzhliu @kevinthesun 

Given that there were previous efforts focused on this. I have done some more comparison

All latencies are in ms.

----------------

CPU - Intel x86 - C5.12x large cascade lake

Case 1 - soiferj@ shape = 1, 12, 128, 128, axis = -1

TVM_NUM_THREADS | Master branch | This PR | Speedup
--|--|--|--
1 | 3.21702 | 3.17762 | 1.0124
2 | 1.60521 | 1.7153 | 0.93582
4 | 0.80477 | 0.98914 | 0.81361
8 | 0.40417 | 0.74009 | 0.54611
16 | 0.4064 | 0.50405 | 0.80626
24 | 0.27334 | 0.50603 | 0.54016

Case 2 - masahi@ shape=(1, 16, 256, 256), axis = 1

TVM_NUM_THREADS | Master branch | This PR | Speedup
--|--|--|--
1 | 2.693 | 2.67314 | 1.00743
2 | 2.69968 | 1.57623 | 1.71274
4 | 2.69336 | 1.0563 | 2.54981
8 | 2.68942 | 0.69915 | 3.84669
16 | 2.68757 | 0.5541 | 4.85034
24 | 2.69724 | 0.43987 | 6.13197

Case 3 - VGG_SSD, shape = (1, 21, 8732), axis = 1

TVM_NUM_THREADS | Master branch | This PR | Speedup
--|--|--|--
1 | 14.81728 | 16.54284 | 0.89569
2 | 14.84842 | 8.9006 | 1.66825
4 | 14.80945 | 5.14822 | 2.87661
8 | 14.78083 | 3.26143 | 4.532
16 | 14.83476 | 2.30334 | 6.44054
24 | 14.86773 | 2.31514 | 6.42196

-----------------

GPU - EC2 G4 instance

Edit - Earlier, I made a mistake in measuring GPU perf.

  | Master branch | This PR | Speedup
-- | -- | -- | --
Case 1 | 0.37284 | 0.03726 | 10.00503
Case 2 | 0.10329 | 0.26443 | 0.39061
Case 3 | 0.02241 | 0.04395 | 0.50989

-------

## Summary
For case 1 on CPU and case 2/3 on GPU, master branch is better performing.

My take on this - We might want to think about design. I think the optimizations that were done for softmax might be possible to do in Relay fusion pass (maybe with some pain).

### CODE REVIEW

**Comment by masahi** [comment]
please make sure you benchmark on 4d spatial inputs. In cuda softmax schedule, there is a special case handling for these inputs.

**Comment by anijain2305** [comment]
> Great job!
> I'm not sure whether we want to keep softmax compute & schedule though. If someone build training on top of relay ir, they might not enable simplifyinference thus need softmax as a single op.

I see. Ok, so I will only change the SimplifyInference pass but keep all the compute/schedule etc. Basically, if one disables SimplifyInference, we should still be able to run

**Comment by anijain2305** [comment]
> please make sure you benchmark on 4d spatial inputs. In cuda softmax schedule, there is a special case handling for these inputs.

I think you meant 2D?
https://github.com/apache/incubator-tvm/blob/master/topi/python/topi/cuda/softmax.py#L55

I see specialization for 2D softmax

**Comment by zhiics** [comment]
I think we may need to keep Softmax schedule as well. We can remove only if we treat them as batchnorm.

But if ones disables simplifyinference, we may still not be able to run if it has batchnorm

**Comment by masahi** [comment]
Correct, but look closer, it is for input dim more than 2d.

See the PR below for background. 
https://github.com/apache/incubator-tvm/pull/2338

**Comment by anijain2305** [comment]
I updated the original post to share more numbers. Please have a look.

@soiferj @masahi

**Comment by anijain2305** [comment]
Incase, anybody wants to run on their end

~~~

import tvm
from tvm import relay
import numpy as np
from tvm.contrib import graph_runtime
from tvm.contrib.debugger import debug_runtime
import time
import os

def test(dshape, axis_):
    dtype = 'float32'
    x = relay.var('x', shape=dshape, dtype=dtype)
    y = relay.nn.softmax(x, axis=axis_)
    func = relay.Function([x], y)
    mod = tvm.relay.Module.from_expr(func)

    np.random.seed(0)
    data = np.random.uniform(size=dshape)

    def run(graph, lib):
        ctx = tvm.gpu()
        module = graph_runtime.create(graph, lib, ctx)
        # module = debug_runtime.create(graph, lib, ctx)
        module.set_input('x', tvm.nd.array(data.astype(dtype)))

        for i in range(0, 50):
            module.run()
        time1 = time.time()

        num_iters = 100
        for i in range(0, num_iters):
            module.run()
        time2 = time.time()
        return (time2 - time1)/num_iters * 1000

    with relay.build_config(opt_level=3, disabled_pass=["SimplifyInference"]):
        graph_master, lib_master, params = relay.build(mod, target="cuda")

    with relay.build_config(opt_level=3):
        graph_new, lib_new, params = relay.build(mod, target="cuda")

    master = run(graph_master, lib_master)
    new_pr = run(graph_new, lib_new)
    print(master, new_pr, master/new_pr, sep='\t', end='\t')

# Soiferj
test((1, 12, 128, 128), -1)

# Masahi
test((1, 16, 256, 256), 1)

# Animesh - VGG SSD
test((1, 21, 8732), 1)
print()
~~~

**Comment by icemelon9** [comment]
Posted a comment on the discussion board. I think we should discuss more on whether we breakdown softmax to a seq of ops.
